merlion/models/layers.py

```diff
@@ -0,0 +1,253 @@
+:py:mod:`AutoML models <merlion.models.automl>`_.
+"""
+from copy import deepcopy
+from typing import Any, Dict, Union
+
+from merlion.models.base import Config, ModelBase
+from merlion.models.factory import ModelFactory
+from merlion.models.anomaly.base import DetectorBase, DetectorConfig
+from merlion.models.forecast.base import ForecasterBase, ForecasterConfig
+from merlion.models.anomaly.forecast_based.base import ForecastingDetectorBase
+from merlion.utils import TimeSeries
+from merlion.utils.misc import AutodocABCMeta
+
+
+class LayeredModelConfig(Config):
+    """
+    Config object for a `LayeredModel`. See `LayeredModel` documentation for more details.
+    """
+
+    def __init__(self, model: Union[ModelBase, Dict], model_kwargs=None, **kwargs):
+        """
+        :param model: The model being wrapped, or a dict representing it.
+        :param model_kwargs: Keyword arguments used specifically to initialize the underlying model. Only used if
+            ``model`` is a dict. Will override keys in the ``model`` dict if specified.
+        :param kwargs: Any other keyword arguments (e.g. for initializing a base class). If ``model`` is a dict,
+            we will also try to pass these arguments when creating the actual underlying model. However, they will
+            not override arguments in either the ``model`` dict or ``model_kwargs`` dict.
+        """
+        # Model-specific kwargs override kwargs when creating the model.
+        model_kwargs = {} if model_kwargs is None else model_kwargs
+        if isinstance(model, dict):
+            model.update({k: v for k, v in kwargs.items() if k not in model and k not in model_kwargs})
+            model, extra_kwargs = ModelFactory.create(**{**model, **model_kwargs, "return_unused_kwargs": True})
+            kwargs.update(extra_kwargs)
+        self.model = model
+        super().__init__(**kwargs)
+
+        # If no model was created, reserve unused kwargs to try initializing the model with
+        if model is None:
+            extra_kwargs = {k: v for k, v in kwargs.items() if k not in self.to_dict()}
+            self.model_kwargs = {**extra_kwargs, **model_kwargs}
+
+    @property
+    def model(self):
+        """
+        The underlying model which this layer wraps.
+        """
+        return self._model
+
+    @model.setter
+    def model(self, model):
+        self._model = model
+        self.model_kwargs = {}
+
+    @property
+    def depth(self):
+        """
+        The depth of the layered model, i.e. the number of layers until we hit a base model. Base models have depth 0.
+        """
+        return 0 if self.model is None else self.model.config.depth + 1
+
+    @property
+    def base_model(self):
+        """
+        The base model at the heart of the full layered model.
+        """
+        model = self.model
+        for _ in range(self.depth - 1):
+            model = model.model
+        return model
```

Comment:
I feel like there are a number of problems with defining a `base_model` method:
1. If you think about wrapper models, they are structured in a tree-like manner. The user-defined model at the very top is the root of the tree, and each underlying model has 1 or more children. Therefore `base_model` is not a well-defined concept (consider a model like `Layer -> Ensemble (Model1, Layer2, Model3) (Layer2 -> Model2)`. While not possible in the current version of Merlion, there is nothing in the Layer code that prevents this from occurring.
2. If we do commit to a linked-list like structure for Layered models, we should be able to either traverse the `model.model` tree all the way down to the base or use `depth` to directly lookup the "`depth`-th" element of the "model array".

I would recommend removing both the `depth` and `base_model` as config parameters for the reasons above.

Reply:
I would argue that the wrapper models are structured in a linked list manner. While we don't explicitly permit looking up the "`depth`-th" element of the linked list as you suggest, there are a number of places where this traversal is useful.

For the time being, I consider Ensembles to themselves be a base model. I think there are some important changes to make to ensembles (some of which are similar to the changes made to Layered Models), but I feel like there are enough differences that I don't want to unify the interfaces just yet. Also, I think implementing full tree-structured models will be a fairly lengthy process that's beyond the scope of this PR.